### PR TITLE
refactoring segmentations with DFAsssembly

### PR DESCRIPTION
**Pull request by DamienGilliard**
image
This PR concerns a small refactoring on the DFJoint_Segmentator component, to use DFAssembly instead of joint meshes and indexes passed separately. Sorry it took so long...
I also added a small runtime warning message is not all the joints could be segmented, for example if the scan did not cover completely the model.

### CODE REVIEW

**Comment by DamienGilliard** [comment]
@9and3  I  believe this PR is ready for review

**Review by 9and3 — COMMENTED** [review]
Hello @DamienGilliard ! Thanks for the PR! 👐 I left a couple of comments and questions, let me know!

File: src/gh/components/DF_joint_segmentator/code.py [hunk 1]
```diff
@@ -24,43 +28,42 @@ def RunScript(self,
...
         if not isinstance(i_clusters[0], Rhino.Geometry.PointCloud):
             raise ValueError("The input clusters must be PointClouds.")
         
-        if not isinstance(i_joints[0], Rhino.Geometry.Mesh):
+        all_joints = i_assembly.all_joints
+        if not isinstance(all_joints[0].faces[0].to_mesh(), Rhino.Geometry.Mesh):
             raise ValueError("The input joints must be convertible to Meshes.")
             
+        # get number of joints
+        n_joints = max([joint.id for joint in all_joints]) + 1
```
[the remark this answers is not in the document]
Comment (DamienGilliard) [reply src/gh/components/DF_joint_segmentator/code.py:49]: Got it. done

File: src/gh/components/DF_joint_segmentator/code.py [hunk 2]
```diff
@@ -7,15 +7,19 @@
...
+    #     super(DFJointSegmentator, self).__init__()
     def RunScript(self, 
-                  i_clusters: Rhino.Geometry.PointCloud, 
-                  i_joints: Rhino.Geometry.Mesh, 
-                  i_joint_ids: int,
+                  i_clusters: typing.List[Rhino.Geometry.PointCloud], 
+                  i_assembly: diffCheck.df_geometries.DFAssembly,
                   i_angle_threshold: float,
                   i_distance_threshold: float):
         """
```
[the remark this answers is not in the document]
Comment (DamienGilliard) [reply src/gh/components/DF_joint_segmentator/code.py:25]: ok !

File: src/gh/components/DF_joint_segmentator/code.py [hunk 3]
```diff
@@ -24,43 +28,42 @@ def RunScript(self,
...
         if len(i_clusters) == 0:
             raise ValueError("No clusters given.")
 
         if not isinstance(i_clusters[0], Rhino.Geometry.PointCloud):
             raise ValueError("The input clusters must be PointClouds.")
         
-        if not isinstance(i_joints[0], Rhino.Geometry.Mesh):
+        all_joints = i_assembly.all_joints
+        if not isinstance(all_joints[0].faces[0].to_mesh(), Rhino.Geometry.Mesh):
             raise ValueError("The input joints must be convertible to Meshes.")
```
[the remark this answers is not in the document]
Comment (DamienGilliard) [reply src/gh/components/DF_joint_segmentator/code.py:46]: Actually it had some sense when using i_joints and i_joint_ids, but not anymore. Was deleted

File: src/gh/components/DF_joint_segmentator/code.py [hunk 4]
```diff
@@ -24,43 +28,42 @@ def RunScript(self,
...
+        n_joints = max([joint.id for joint in all_joints]) + 1
 
         # prepping the reference meshes
-        n_joints = max(i_joint_ids) + 1
-        joints = [ [] for i in range(n_joints) ]
-        for face, id in zip(i_joints, i_joint_ids):
-            face.Subdivide()
-            face.Faces.ConvertQuadsToTriangles()
-            joints[id].append(df_cvt.cvt_rhmesh_2_dfmesh(face))
+        joints = [[] for _ in range(n_joints)]
```
[the remark this answers is not in the document]
Comment (DamienGilliard) [reply src/gh/components/DF_joint_segmentator/code.py:52]: indeed, it did it as well with other variables

File: src/gh/components/DF_joint_segmentator/code.py [hunk 5]
```diff
@@ -72,19 +75,29 @@ def RunScript(self,
...
+        o_reference_point_clouds = []
+        for  joint_segment, transform, _joint_cloud in zip(joint_segments, transforms, joint_clouds):
+            if joint_segment.IsValid:
+                o_joint_segments.append(joint_segment)
+                o_transforms.append(transform)
+                o_reference_point_clouds.append(_joint_cloud)
+            else:
+                ghenv.Component.AddRuntimeMessage(RML.Warning, "Some joints could not be segmented and were ignored.")
+
+        return o_joint_segments, o_transforms, o_reference_point_clouds
```
[the remark this answers is not in the document]
Comment (DamienGilliard) [reply src/gh/components/DF_joint_segmentator/code.py:97]: a) No, right now it sifts and removes invalid point clouds. What we can do is leave the invalid ones (that are converted into None) as you suggest and check at the visualization step if it is None, and if so color it black or white, or another way to signal "Hey, there was no segment for this joint".  I believe we said last time that we should have only "evaluable" data outputted, but I think this makes more sense.  Do you agree @eleniv3d  @mariklad  ? And anyway, if we remove elements, we have to know where they were removed so the evaluation step compares the right point cloud, so it is better to do it at the visualization step.  

b) Indeed, we had decided to let down the point to point evaluation, in favor of point to mesh, so this is not needed anymore.

File: src/gh/diffCheck/diffCheck/df_geometries.py [hunk 6]
```diff
@@ -422,18 +422,21 @@ def dump_xml(self, pretty_xml: str, dir: str):
 
     @property
     def all_joint_faces(self):
-        for beam in self.beams:
-            self._all_jointfaces.extend(beam.joint_faces)
+        if len(self._all_jointfaces) == 0:
+            for beam in self.beams:
+                self._all_jointfaces.extend(beam.joint_faces)
```
[the remark this answers is not in the document]
Comment (DamienGilliard) [reply src/gh/diffCheck/diffCheck/df_geometries.py:427]: Indeed. 
What I did to improve it was set the list empty before, and remove the `if len()==0` . That way if the DFAssembly is augmented by one or many beams, the value is re-calculated at the moment you re-access the property:

image

File: src/gh/components/DF_joint_segmentator/code.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (9and3) [reply src/gh/components/DF_joint_segmentator/code.py]: ok marked in issue tracking for the PR for #50

File: src/gh/diffCheck/diffCheck/df_geometries.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (9and3) [reply src/gh/diffCheck/diffCheck/df_geometries.py]: better! Nice solution 🍰

**Comment by 9and3** [comment]
pop a message @DamienGilliard when you are ready for a merge on this! 👐

**Comment by DamienGilliard** [comment]
@9and3 , Went through it again and I believe it is ready ! Your comments were addressed.

**Review by 9and3 — APPROVED** [review]
lgtm!

**Pull request merged**